calico_kubernetes/policy.py

```diff
@@ -0,0 +1,224 @@
+# Copyright 2015 Metaswitch Networks
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Usage: parse (allow|deny) [(
+      (tcp|udp) [(from [(ports <SRCPORTS>)] [(label <SRCLABEL>)] [(cidr <SRCCIDR>)])]
+                [(to   [(ports <DSTPORTS>)] [(label <DSTLABEL>)] [(cidr <DSTCIDR>)])] |
+      icmp [(type <ICMPTYPE> [(code <ICMPCODE>)])]
+           [(from [(label <SRCLABEL>)] [(cidr <SRCCIDR>)])]
+           [(to   [(label <DSTLABEL>)] [(cidr <DSTCIDR>)])] |
+      [(from [(label <SRCLABEL>)] [(cidr <SRCCIDR>)])]
+      [(to   [(label <DSTLABEL>)] [(cidr <DSTCIDR>)])]
+    )]
+"""
+import re
+import logging
+from netaddr import IPNetwork, AddrFormatError
+from docopt import docopt
+from pycalico.datastore_datatypes import Rule
+
+_log = logging.getLogger(__name__)
+
+
+class PolicyParser(object):
+    """
+    Class which parses Kubernetes annotation-based policy strings and
+    converts them to libcalico Rules objects.
+
+    Usage:
+      parser = PolicyParser(namespace)
+      rule = parser.parse_line("allow tcp from label stage=prod")
+    """
+    def __init__(self, namespace):
+        self.namespace = namespace
+
+    def parse_line(self, policy):
+        """
+        Takes a single line of policy as defined in the annotations of a
+        pod and returns the equivalent libcalico Rule object.
+        :param policy: Policy string from annotations.
+        :return: A Rule object which represent the given policy.
+        """
+        _log.info("Parsing policy line: '%s'", policy)
+        splits = policy.split()
+
+        try:
+            args = docopt(__doc__, argv=splits)
+        except BaseException:
+            raise PolicyParseError("Failed to parse policy: %s", policy)
+
+        # Generate a rule object from the arguments.
+        rule = self._generate_rule(args)
+
+        return rule
+
+    def _generate_rule(self, arguments):
+        """
+        Generates a libcalico
+        :param arguments: A dictionary of arguments with the following format.
+
+        {
+            "<DSTCIDR>": null,
```

Comment:
I'd be more inclined to make this function's docstring contain the docopt usage (if possible), or make the docopt usage string a variable inside this function, rather than duplicate the text.